libs/entitlement/asm-provider/src/main/java/org/elasticsearch/entitlement/instrumentation/impl/InstrumentationServiceImpl.java

```diff
@@ -20,47 +20,74 @@
+
     @Override
     public Instrumenter newInstrumenter(Class<?> clazz, Map<MethodKey, CheckMethod> methods) {
         return InstrumenterImpl.create(clazz, methods);
     }
 
     @Override
     public Map<MethodKey, CheckMethod> lookupMethods(Class<?> checkerClass) throws IOException {
-        var methodsToInstrument = new HashMap<MethodKey, CheckMethod>();
-        var classFileInfo = InstrumenterImpl.getClassFileInfo(checkerClass);
-        ClassReader reader = new ClassReader(classFileInfo.bytecodes());
-        ClassVisitor visitor = new ClassVisitor(Opcodes.ASM9) {
-            @Override
-            public MethodVisitor visitMethod(
-                int access,
-                String checkerMethodName,
-                String checkerMethodDescriptor,
-                String signature,
-                String[] exceptions
-            ) {
-                var mv = super.visitMethod(access, checkerMethodName, checkerMethodDescriptor, signature, exceptions);
-
-                var checkerMethodArgumentTypes = Type.getArgumentTypes(checkerMethodDescriptor);
-                var methodToInstrument = parseCheckerMethodSignature(checkerMethodName, checkerMethodArgumentTypes);
-
-                var checkerParameterDescriptors = Arrays.stream(checkerMethodArgumentTypes).map(Type::getDescriptor).toList();
-                var checkMethod = new CheckMethod(Type.getInternalName(checkerClass), checkerMethodName, checkerParameterDescriptors);
-
-                methodsToInstrument.put(methodToInstrument, checkMethod);
-
-                return mv;
-            }
-        };
-        reader.accept(visitor, 0);
+        Map<MethodKey, CheckMethod> methodsToInstrument = new HashMap<>();
+
+        ArrayDeque<Class<?>> classesToVisit = new ArrayDeque<>(Collections.singleton(checkerClass));
+        while (classesToVisit.isEmpty() == false) {
+            var currentClass = classesToVisit.remove();
+
+            var classFileInfo = InstrumenterImpl.getClassFileInfo(currentClass);
+            ClassReader reader = new ClassReader(classFileInfo.bytecodes());
+            ClassVisitor visitor = new ClassVisitor(Opcodes.ASM9) {
+
+                @Override
+                public void visit(int version, int access, String name, String signature, String superName, String[] interfaces) {
+                    super.visit(version, access, name, signature, superName, interfaces);
+                    try {
+                        if (OBJECT_INTERNAL_NAME.equals(superName) == false) {
+                            var className = Type.getObjectType(superName).getClassName();
+                            classesToVisit.add(Class.forName(className));
+                        }
+                        for (var interfaceName : interfaces) {
+                            var className = Type.getObjectType(interfaceName).getClassName();
+                            classesToVisit.add(Class.forName(className));
```

Comment:
Note that this kind of logic can cause us to re-visit interfaces in cases of repeated inheritance. In general, we ought to visit each interface just once.

In fact, this logic could re-visit some interfaces an exponential number of times! Consider this inheritance hierarchy:

```
  EightTimes
     /  \
    A    B
     \  /
  FourTimes
     /  \
    C    D
     \  /
   TwoTimes
     /  \
    E    F
     \  /
 StartingClass
```

Reply:
True; we could visit in 2 passes, or be aware of this and don't care. It's for a very limited scenario we control, not general code.

Reply:
You don't need two passes, just a checklist (say `HashSet`) of interfaces you've already added, or a queue data structure that rejects dupes (again, `HashSet` works, though it's inconvenient to pop an arbitrary item from that).